### ISSUE TITLE
AuthorizationUrl missing parameters on Unity 2020 UWP

### ISSUE THREAD

**Issue by Maesla** [cdmvision/authentication-unity#8]
AuthorizationCodeFlow::GetAuthorizationUrl() is missing parameters when the application is deployed to UWP using Unity 2020

### Step to reproduce

1. Deploy to a UWP application using Unity 2020.3.41
2. See AuthorizationCodeFlow::AuthorizationUrl()
3. Check that several parameters are missing, such as required_type and client_id

### Causes
GetAuthorizationUrl() uses JsonHelper.ToDictionary.
ToDictionary uses JObject.FromObject(obj).ToObject<Dictionary<string, string>>();
JObject.FromObject(obj) somehow doesn't work properly on UWP 2020.3.41, skipping some parameters

### Remarks
JObject.FromObject(obj) skips some paremeters, so they are not added to the dictionary, and, therefore, into the parameters.
This behaviour is very specific, because I have been able to reproduce only on Unity 2020.3.41 UWP when the class is into the package folder, so I guess that has to be something about getting the fields by reflection when the class is in the packages

### Workarrounds and fixes

- Unity 2021 seems to work
- Move the code outside the package folder
- Override JsonHelper.ToDictionary to force building the dictionary correctly
```csharp
        public static Dictionary<string, string> ToDictionary(object obj)
        {
            var dictionary = new Dictionary<string, string>();

            if (obj is AuthorizationCodeRequest authorizationCodeRequest)
            {
                dictionary["response_type"] = authorizationCodeRequest.responseType;
                dictionary["client_id"] = authorizationCodeRequest.clientId;
                dictionary["redirect_uri"] = authorizationCodeRequest.redirectUri;
                dictionary["scope"] = authorizationCodeRequest.scope;
                dictionary["state"] = authorizationCodeRequest.state;
            }
            else if (obj is AccessTokenRequest accessTokenRequest)
            {
                dictionary["grant_type"] = "authorization_code";
                dictionary["code"] = accessTokenRequest.code;
                dictionary["client_id"] = accessTokenRequest.clientId;
                dictionary["client_secret"] = accessTokenRequest.clientSecret;
                dictionary["redirect_uri"] = accessTokenRequest.redirectUri;
            }
            else if (obj is RefreshTokenRequest refreshToken)
            {
                dictionary["grant_type"] = "refresh_token";
                dictionary["refresh_token"] = refreshToken.refreshToken;
                dictionary["scope"] = refreshToken.scope;
            }
            else
            {
                dictionary = JObject.FromObject(obj).ToObject<Dictionary<string, string>>();
            }

            if (dictionary != null)
            {
                // Remove empty parameters.
                var keys = dictionary.Keys.Where(key => string.IsNullOrEmpty(dictionary[key])).ToArray();
                foreach (var key in keys)
                {
                    dictionary.Remove(key);
                }
            }

            return dictionary;
        }
```

### Test and screenshot
I've created a test project with this issue.
The serializable class is on a custom package inside package folder
The test code is this:
```csharp
using Newtonsoft.Json.Linq;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using UnityEngine;
using UnityEngine.Assertions;
using UnityEngine.UI;

public class TestDataContract : MonoBehaviour
{
    public InputField inputField;
    private void Start()
    {
        var request = new AuthorizationCodeRequest()
        {
            clientId = "john doe",
            redirectUri  = "http://localhost"
        };

        var dictionary = ToDictionary(request);

        StringBuilder builder = new StringBuilder();
        builder.AppendLine($"Dictionary Count: {dictionary.Count}");

        foreach (var kvp in dictionary)
        {
            builder.AppendLine($"Key: {kvp.Key} Value: {kvp.Value}");
        }

        inputField.text = builder.ToString();
    }

    public static Dictionary<string, string> ToDictionary(object obj)
    {
        var dictionary = JObject.FromObject(obj).ToObject<Dictionary<string, string>>();

        if (dictionary != null)
        {
            // Remove empty parameters.
            var keys = dictionary.Keys.Where(key => string.IsNullOrEmpty(dictionary[key])).ToArray();
            foreach (var key in keys)
            {
                dictionary.Remove(key);
            }
        }

        return dictionary;
    }
}
```

On Editor:
![image](https://user-images.githubusercontent.com/25863696/217539422-60ff68e6-e7eb-47ef-a047-464d84918318.png)

On UWP
![image](https://user-images.githubusercontent.com/25863696/217539468-cabd75b4-9e9d-4397-b009-e1fdf204d9e2.png)

**Comment by Maesla** [comment]
@ibrahimpenekli the bug is so weird and so specific that I don't expect any fix, but I thought that it could be helpful to someone else in the same situation, so please feel free to close it

**Comment by Maesla** [comment]
I reported the bug to Unity and they've confirmed and apparently it's also present on Unity 2023

https://issuetracker.unity3d.com/issues/newtonsoft-json-doesnt-correctly-serialize-classes-inside-the-packages-folder-when-in-uwp-and-webgl-builds

**Comment by ibrahimpenekli** [comment]
Hi @Maesla ,

Thanks for the bug report.
Did you try yo add [Preserve] attribute to the DataContracts?

**Comment by Maesla** [comment]
You mean this right?
[assembly: Preserve]
https://docs.unity3d.com/2018.3/Documentation/Manual/ManagedCodeStripping.html

I've just tried, same result

**Comment by ibrahimpenekli** [comment]
Yes it was, did you try to add [Preserve] attribute to [DataMember]s also?

**Comment by Maesla** [comment]
That did the trick!

```csharp
using System.Runtime.Serialization;
using UnityEngine.Scripting;

/// </summary>
[DataContract]
public class AuthorizationCodeRequest
{
    [DataMember(Name = "response_type", IsRequired = true)]
    [Preserve]
    public string responseType => "code"; //Correctly Serialized

    [DataMember(Name = "client_id", IsRequired = true)]
    public string clientId { get; set; } //Not serialized

    [DataMember(Name = "redirect_uri")]
    [Preserve]
    public string redirectUri { get; set; } //Correctly Serialized
}
```
![image](https://user-images.githubusercontent.com/25863696/220087350-edb0a644-0405-421b-ba53-8296f21b029d.png)

I checked in WebGL but I suppose the behavior in UWP should be the same

**Comment by ibrahimpenekli** [comment]
I'm glad to see it worked! Normally, Unity should not strip out these fields that are used in the application. We may include [Preserve] for each serializable data to avoid this bug.

Will you make a PR or should I mark them with [Preserve]?

**Comment by Maesla** [comment]
Yes, it should be a bug in Unity Linker, because it should take those into account. Actually, it takes them into account for some compilations, such as Desktop, and in some Unity Versions, such as 2021
https://docs.unity3d.com/Manual/unity-linker.html

btw all the tests were done with stripping level at LOW
Please, you go ahead marking the fields as Preserve, because you know better than me which classes and fields must be serialized.

**Comment by ibrahimpenekli** [comment]
Fixed in 1f2b2ab
